Doc/library/asyncio-sync.rst

```diff
@@ -1,305 +1,324 @@
 .. currentmodule:: asyncio
+
 .. _asyncio-sync:
 
-Synchronization primitives
 ==========================
+Synchronization Primitives
+==========================
+
+asyncio synchronization primitives are designed to be similar to
+those of the :mod:`threading` module with two important caveats:
+
+* asyncio primitives are not thread-safe, therefore they should not
+  be used for OS threads synchronization (use :mod:`threading` for
+  that);
 
-**Source code:** :source:`Lib/asyncio/locks.py`
+* methods of synchronization objects do not accept the *timeout*
+  argument; use the :func:`asyncio.wait_for` function to perform
+  operations with timeouts.
 
-Locks:
+asyncio has the following basic primitives:
 
 * :class:`Lock`
 * :class:`Event`
 * :class:`Condition`
-
-Semaphores:
-
 * :class:`Semaphore`
 * :class:`BoundedSemaphore`
 
-asyncio lock API was designed to be close to classes of the :mod:`threading`
-module (:class:`~threading.Lock`, :class:`~threading.Event`,
-:class:`~threading.Condition`, :class:`~threading.Semaphore`,
-:class:`~threading.BoundedSemaphore`), but it has no *timeout* parameter. The
-:func:`asyncio.wait_for` function can be used to cancel a task after a timeout.
+
+---------
 
 
 Lock
-----
+====
 
 .. class:: Lock(\*, loop=None)
 
-   Primitive lock objects.
+   Implements a mutex lock for asyncio tasks.  Not thread-safe.
 
-   A primitive lock is a synchronization primitive that is not owned by a
-   particular coroutine when locked.  A primitive lock is in one of two states,
-   'locked' or 'unlocked'.
+   An asyncio lock can be used to guarantee exclusive access to a
+   shared resource to a single asyncio Task.
 
-   The lock is created in the unlocked state.
-   It has two basic methods, :meth:`acquire` and :meth:`release`.
-   When the state is unlocked, acquire() changes the state to
-   locked and returns immediately.  When the state is locked, acquire() blocks
-   until a call to release() in another coroutine changes it to unlocked, then
-   the acquire() call resets it to locked and returns.  The release() method
-   should only be called in the locked state; it changes the state to unlocked
-   and returns immediately.  If an attempt is made to release an unlocked lock,
-   a :exc:`RuntimeError` will be raised.
+   The preferred way to use a Lock is an :keyword:`async with`
```

Comment:
Perhaps have the example code after introducing: acquire, lock, release, locked.

Reply:
Yeah, I thought about it too.  If we do that the examples won't be immediately visible to the reader.  OTOH acquire and release are so wide known that understanding these short examples (read "usage recommendations") shouldn't be a problem...

Reply:
Good point.